File: day02/puzzle.py
from aoc import AocPuzzle


TEST_ANSWERS = {
    "part1": 150,
    "part2": 900
}


class Command:
    def __init__(self, cmd: str):
        self.direction, units = cmd.split(" ")
        self.units = int(units)


class Puzzle(AocPuzzle[list[Command], int]):
    def __init__(self):
        super().__init__(TEST_ANSWERS)


    def parse_input(self, lines: list[str]):
        return [Command(line) for line in lines]


    def run_part1(self):
        horizontal = 0
        depth = 0
        for command in self._input:
            match command.direction:
                case "forward":
                    horizontal += command.units
                case "down":
                    depth += command.units
                case "up":
                    depth -= command.units

        self.print_position(horizontal, depth)
        return horizontal * depth


    def run_part2(self):
        horizontal = 0
        depth = 0
        aim = 0
        for command in self._input:
            match command.direction:
                case "forward":
                    horizontal += command.units
                    depth += aim * command.units
                case "down":
                    aim += command.units
                case "up":
                    aim -= command.units

        self.print_position(horizontal, depth)
        return horizontal * depth


    def print_position(self, horizontal: int, depth: int):
        print("Horizontal:", horizontal)
        print("Depth:", depth)

File: day19/Puzzle.cs
using System.Collections.Generic;
using System.Linq;

namespace Aoc2021
{
    public class Puzzle : AocPuzzle<Scanner[], int>
    {
        protected override IDictionary<string, int> TestAnswers { get; } = new Dictionary<string, int>
        {
            { "part1", 79 },
            { "part2", 3621 }
        };

        protected override Scanner[] ParseInput(string[] lines)
        {
            var scanners = new List<Scanner>();
            var beacons = new List<Vector>();
            foreach (string line in lines)
            {
                if (line.Length == 0)
                {
                    scanners.Add(new Scanner(beacons.ToArray()));
                    continue;
                }

                if (line.StartsWith("---"))
                {
                    beacons.Clear();
                    continue;
                }

                var coords = line.Split(',').Select(coord => int.Parse(coord)).ToArray();
                beacons.Add(new Vector(coords[0], coords[1], coords[2]));
            }

            scanners.Add(new Scanner(beacons.ToArray()));
            return scanners.ToArray();
        }

        protected override int RunPart1()
        {
            var beacons = FindBeacons();
            return beacons.Count;
        }

        protected override int RunPart2()
        {
            FindBeacons();
            int maxDistance = 0;
            for (int i = 0; i < _input.Length - 1; ++i)
            {
                for (int j = i + 1; j < _input.Length; ++j)
                {
                    var distance = _input[i].Position - _input[j].Position;
                    int manhattanDistance = distance.ManhattanLength;
                    if (manhattanDistance > maxDistance)
                        maxDistance = manhattanDistance;
                }
            }

            return maxDistance;
        }

        private HashSet<Vector> FindBeacons()
        {
            var scanners = _input[1..].ToList();
            var processedScanners = new List<Scanner> { _input[0] };

            while (scanners.Count > 0)
            {
                for (int i = 0; i < processedScanners.Count; ++i)
                {
                    for (int j = 0; j < scanners.Count; ++j)
                    {
                        for (int orientation = 0; orientation < 24; ++orientation)
                        {
                            var beacons = scanners[j].GetBeacons(orientation);
                            var offset = processedScanners[i].GetBeaconOffset(beacons, 12);
                            if (offset.HasValue)
                            {
                                scanners[j].Position = processedScanners[i].Position + offset.Value;
                                scanners[j].Beacons = beacons;
                                processedScanners.Add(scanners[j]);
                                scanners.RemoveAt(j--);
                                break;
                            }
                        }
                    }
                }
            }

            var foundBeacons = new HashSet<Vector>();
            foreach (var scanner in processedScanners)
            {
                foundBeacons.UnionWith(scanner.Beacons.Select(b => b + scanner.Position));
            }

            return foundBeacons;
        }
    }
}